articles/fork-in-C-programming-language/README.md

```diff
@@ -0,0 +1,205 @@
+# fork() in C programming language
+
+![hero-image](/engineering-education/fork-in-C-programming-language/hero.png)
+
+### Introduction
+You have seen a lot of processes in your task manager if you are using Windows. And in your resource monitor if you are using Linux. But did you think even once how these are created? In this tutorial, we will talk about fork then implement some examples in the C programming language. 
+
+### Prerequisites
+To follow along with this tutorial, you should have:
+- A good understanding of C programming language.
+- Familiar with Unix-like operating systems.
+
+### What is Fork()?
+In the computing field, **Fork** is the primary method of process creation on Unix-like operating systems. Which creates a new copy called *child* out of the original process that called *parent*. And when the parent process is closed or crashed for some reason, it also kills the child process. 
+
+Let's start with the life-cycle of a process:
+
+![Process life-cycle](/engineering-education/fork-in-C-programming-language/ProcessState.jpg)
+
+[Image Source](https://www.cs.uic.edu/~jbell/CourseNotes/OperatingSystems/3_Processes.html).
+
+The operating system is using a unique id for every process to keep track of all processes. And for that, fork() doesn't take any parameter and return an int value as following:
+ - Zero: if it is the child process(the process created).
+ - Positive value: if it is the parent process.
+ - Negative value: if an error occurred.
+ 
+*Note*: The following codes only run in Linux and UNIX based. If you are running Windows, then I recommend you to use [Cygwin](https://www.cygwin.com/). 
+
+Let's jump into the practical section where we will create examples from the simple level to the advanced one.
+
+### Hello world!
+```c
+#include <stdio.h> 
+#include <sys/types.h> 
+#include <unistd.h> 
+int main() 
+{ 
+    fork(); 
+    printf("Hello world!\n"); 
+    return 0; 
+} 
+```
+
+The output will be:
+```
+Hello world!
+Hello world!
+```
+Where one of the output came from the parent process and the other one from the child process.
+
+![Simple fork](/engineering-education/fork-in-C-programming-language/fork.png)
+
+Simply, we can tell that the result is 2 power of n, where n is the number of fork() system calls. For example:
+```c
+#include <stdio.h> 
+#include <sys/types.h> 
+#include <unistd.h> 
+#include <stdlib.h>
+int main() 
+{ 
+    fork(); 
+    fork(); 
+    fork(); 
+    printf("Hello world!\n");
+    return 0; 
+} 
+```
+The result is:
+```
+Hello world!
+Hello world!
+Hello world!
+Hello world!
+Hello world!
+Hello world!
+Hello world!
+Hello world!
+```
+Another example is:
+```c
+int main() {
+  if(fork() ==0)
+    if(fork())
+      printf("Hello world!!\n");
+  exit(0);
+  }
+```
+I drew a brief sketch to let you understand the idea:
+
+![Fork](/engineering-education/fork-in-C-programming-language/fork1.png)
+
```

Comment:
A beginner cannot understand this diagram, unless you give a detailed explanation. Please add one :)